### DISCUSSION TITLE
pgvector & pg-storm integration

### DISCUSSION THREAD

**Discussion by CC-Hsu** [heterodb/pg-strom#780] [Ideas / Features]
Hi, Team,

It is now really popular on LLM and embedding vector feature support for databases, especially [the post popular `pgvector` extensoin for PGSQL](https://github.com/pgvector/pgvector).

However, [this extension is not going to be GPU-accelerated](https://github.com/pgvector/pgvector/issues/1#issuecomment-969792433).

I think it would be great if PG-Storm can support `pgvector` integration and acceleration (maybe walk together).

Specifically, I think PG-Storm can toward in-depth GPU-accelerated support on
1. `vector` Datatype support
2. Function support, such as distance functions, comparison operators
3. Index support (if possible), including searching index, (re)building index

I feel it would be popular if such feature is available.

Best Regards,

**Comment by kaigai** [comment]
I don't know whether "pgvector integration" is a suitable direction, or not.
However, I have worked on non-index based vector search with PG-Strom, and probably it makes sense for the recent demands.
We already list up items to be developed.
- #698 vector search functionality (GPU version of distance function based on float2/4/8[] array, or harhaps `vector` type)
- #738 GpuSort + Limit (A feature to pick up the nearest N items from the query results)
- #752 Large-tables GpuJoin -> it also allows to preserve GpuScan results on GPU memory to reuse other task's input like GpuJoin and potentialy GpuSort.

I plan to implement these features by end of this summer.
If you have real-life vector search workloads to test these workloads, it is welcome for us.
